### DISCUSSION TITLE
VisUI: support FileChooser

### DISCUSSION THREAD

**Discussion by ishitatsuyuki** [libktx/ktx#103] [Ideas]
While it's a desktop only widget, I think it worth to add support for a builder for it.

**Comment by czyzby** [comment]
I think there were some issues with including desktop-only utilities in the core lib, but I honestly don't remember why we decided against it back then. @kotcrab Can `FileChooser` DSL cause any problems with Android/iOS builds?

**Comment by kotcrab** [comment]
I don't remember either but I don't think so. I moved usage of all problematic APIs (`java.awt` and `sun.awt`)  behind reflection so should be fine on Android but I can't say anything about iOS. 

It's also a question whether should we have builder method that works only on one platform with all the other cross-platform ones.

**Comment by czyzby** [comment]
Now that you mention it, I think these were the original reasons:

- `FileChooser` is limited to a single platform, while all other KTX utilities work everywhere. It would be somewhat inconsistent.
- There's no need for an advanced GUI builder when it comes to `FileChooser`, since its API is pretty much "self-sufficient". A builder for FC would end up a fancy constructor with `apply` equivalent, since you're unlikely to add custom widgets to its `Window`.

I might be wrong though - is there a common use case where you'd want to add actors to a `FileChooser`?

@ishitatsuyuki If you don't agree, and if you're willing to set up a pull request with a `FileChooser` builder prototype, I'll be happy to review it.

**Comment by ishitatsuyuki** [comment]
Yeah, I agree. It would end up as an apply style constructor, and I guess it's even trickier as the documentation suggests to reuse single widget, while others are mainly constructed for every demand.

Feel free to close this issue if this has too much concerns to implement.

**Comment by czyzby** [comment]
I'd be willing to sacrifice a _little_ bit consistency for usability, but I just don't see how `FileChooser` would need access to GUI building code. Feel free to comment it if you find a use case for this utility.
